core/src/main/java/org/elasticsearch/search/aggregations/bucket/adjacency/InternalAdjacencyFilters.java

```diff
@@ -0,0 +1,209 @@
+        List<InternalBucket> buckets = new ArrayList<>(size);
+        for (int i = 0; i < size; i++) {
+            buckets.add(new InternalBucket(in));
+        }
+        this.buckets = buckets;
+        this.bucketMap = null;
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeVInt(buckets.size());
+        for (InternalBucket bucket : buckets) {
+            bucket.writeTo(out);
+        }
+    }
+
+    @Override
+    public String getWriteableName() {
+        return AdjacencyFiltersAggregationBuilder.NAME;
+    }
+
+    @Override
+    public InternalAdjacencyFilters create(List<InternalBucket> buckets) {
+        return new InternalAdjacencyFilters(this.name, buckets, this.pipelineAggregators(), this.metaData);
+    }
+
+    @Override
+    public InternalBucket createBucket(InternalAggregations aggregations, InternalBucket prototype) {
+        return new InternalBucket(prototype.key, prototype.docCount, aggregations);
+    }
+
+    @Override
+    public List<InternalBucket> getBuckets() {
+        return buckets;
+    }
+
+    @Override
+    public InternalBucket getBucketByKey(String key) {
+        if (bucketMap == null) {
+            bucketMap = new HashMap<>(buckets.size());
+            for (InternalBucket bucket : buckets) {
+                bucketMap.put(bucket.getKey(), bucket);
+            }
+        }
+        return bucketMap.get(key);
+    }
+
+    @Override
+    public InternalAggregation doReduce(List<InternalAggregation> aggregations, ReduceContext reduceContext) {
+        Map<String, List<InternalBucket>> bucketsMap = new HashMap<>();
+        for (InternalAggregation aggregation : aggregations) {
+            InternalAdjacencyFilters filters = (InternalAdjacencyFilters) aggregation;
+            for (InternalBucket bucket : filters.buckets) {
+                List<InternalBucket> sameRangeList = bucketsMap.get(bucket.key);
+                if(sameRangeList == null){
+                    sameRangeList = new ArrayList<>(aggregations.size());
+                    bucketsMap.put(bucket.key, sameRangeList);
+                }
+                sameRangeList.add(bucket);
+            }
+        }
+
+        InternalAdjacencyFilters reduced = new InternalAdjacencyFilters(name, 
+                new ArrayList<InternalBucket>(bucketsMap.size()), pipelineAggregators(), 
+                getMetaData());
+        for (List<InternalBucket> sameRangeList : bucketsMap.values()) {
+            reduced.buckets.add((sameRangeList.get(0)).reduce(sameRangeList, reduceContext));
+        }
+        return reduced;
+    }
+
+    @Override
+    public XContentBuilder doXContentBody(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(CommonFields.BUCKETS);
+        for (InternalBucket bucket : buckets) {
+            bucket.toXContent(builder, params);
+        }
+        builder.endObject();
+        return builder;
```

Comment:
You decided to go with the keyed way for the xcontent representation but I am not sure that I like it. For instance, if there are two filters A and B, the user has no way to know whether the key will be `A&B` or `B&A` so these keys would be hard to use? There are also potential corner cases if the filter names use the separator character in their name, which is something unlikely but that I still like to avoid if possible... How about having buckets look like below:

```
buckets: [
  {
    "filters": ["A", "A"],
    "doc_count": 42,
    "aggs": { ... }
  },
  {
    "filters": ["A", "B"],
    "doc_count": 12,
    "aggs": { ... }
  },
  {
    "filters": ["B", "B"],
    "doc_count": 20,
    "aggs": { ... }
  }
]
```